rfcs/0019-storing-client-data.md

```diff
@@ -0,0 +1,1033 @@
+__NOTE: We refer `external_transitions` or transitions as blocks in this RFC__
+
+# Summary
+
+The Coda protocol contains many different types of objects (blocks, transactions, accounts) and these objects have relationships with each other. A client communicating with the protocol would often make queries involving the relationship of these objects (i.e. Find the 10 most recent transactions that Alice sent after sending transactions, TXN). This RFC will discuss how we can take advantage of these relationships by using relational databases. This will accelerate our process in writing concise and maintainable queries that are extremely performant. We will further more discuss the data flow that connects the daemon to these storage components and make efficient queries. We will also discuss the pros and cons of using this new design.
```

Comment:
Can we clean up this summary so people won't be confused about what exactly we're proposing here? There was some confusion by a coworker this morning.

Here's my suggestion:

The Coda protocol contains many different types of objects (blocks, transactions, accounts) and these objects have relationships with each other. A client communicating with a protocol node may wish to make queries involving the relationship of these objects (i.e. Find the 10 most recent transactions that Alice sent after sending transactions, TXN). This RFC will discuss an architecture for optionally including a relational database to be able to answer these queries. Note that the architecture proposed in this document does not couple a node to one particular database, nor even couple it to _a_ database. It actually does the opposite -- moving lots of logic that currently is inside the daemon process into another one that optionally can be omitted when a user doesn't need to use it. 

When a user does want to make such queries, this architecture will accelerate the process of writing concise and maintainable queries that are extremely performant. We will further more discuss the data flow that optionally connects the daemon to these storage components and make efficient queries. We will also discuss the pros and cons of using this new design.

Reply:
Further:

This architecture entails ripping out the API and archive related work into separate processes. By keeping the API separate, it can be scaled in load balanced, as can the database, without having to scale the number of nodes that are speaking to the archiver (unless you want higher network consistency).

Reply:
@bkase, so awesome. I love it